bug-finding.ipynb

```diff
@@ -288,13 +338,24 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Bug 3 analysis and response"
+    "# Bug 3 analysis and response\n",
+    "\n",
+    "The bug occurs when we try to use list comprehension in this problem:\n",
+    "cooccurrences = [1 for l in chunked_lemmata if w1 in l and w2 in l]\n",
+    "\n",
+    "Because of this o11 is calculated incorrectly and overstated (larger value than expected), and subsequently o12 turns out to be negative as it is defined as o12 = r1 - o11, as o11 turns out to be larger than r1.\n",
+    "\n",
+    "By using another list comprehension to return indexes like this:\n",
+    " indexes = [i for i, lemma in enumerate(lemmata) if lemma == w1]\n",
+    "We can check for the presence of w2 in the window where the lemma w1 is contained.\n",
+    "\n",
+    "Using a simple for loop, we can iterate over these indexes and sum the cooccurences - which returns a value for o11, less than r1 - and also makes o12 positive."
```

Comment:
Excellent, and exactly right -- the original implementation could double-count co-occurrences under certain circumstances.